### PR TITLE
Improved status cleanup

### PR DESCRIPTION
**Pull request by georgehrke**

### CODE REVIEW

File: apps/user_status/lib/Connector/UserStatus.php [hunk 1]
```diff
@@ -56,8 +57,8 @@ public function __construct(Db\UserStatus $status) {
 		$this->message = $status->getCustomMessage();
 		$this->icon = $status->getCustomIcon();
 
-		if ($status->getStatus() === 'invisible') {
-			$this->status = 'offline';
+		if ($status->getStatus() === StatusService::INVISIBLE) {
+			$this->status = StatusService::OFFLINE;
```
Comment (nickvergessen) [inline apps/user_status/lib/Connector/UserStatus.php:61]:
```suggestion
		if ($status->getStatus() === IUserStatus::INVISIBLE) {
			$this->status = IUserStatus::OFFLINE;
```

Invisible is missing from the public interface, but I guess it also makes sense

File: apps/user_status/lib/Service/StatusService.php [hunk 2]
```diff
@@ -54,17 +54,38 @@ class StatusService {
...
-		'online',
-		'away',
-		'dnd',
-		'invisible',
-		'offline'
+	public const ONLINE = 'online';
+	public const AWAY = 'away';
+	public const DND = 'dnd';
+	public const INVISIBLE = 'invisible';
+	public const OFFLINE = 'offline';
```
Comment (nickvergessen) [inline apps/user_status/lib/Service/StatusService.php:61]: Those are in `OCP\UserStatus\IUserStatus` apart from invis. I'd say just add invis there too and use them

**Review by nickvergessen — APPROVED** [review]

File: apps/user_status/lib/Db/UserStatusMapper.php [hunk 3]
```diff
@@ -125,12 +126,31 @@ public function findByUserIds(array $userIds):array {
...
+	 */
+	public function clearStatusesOlderThan(int $olderThan, int $now): void {
+		$qb = $this->db->getQueryBuilder();
+		$qb->update($this->tableName)
+			->set('status', $qb->createNamedParameter(IUserStatus::OFFLINE))
+			->set('is_user_defined', $qb->createNamedParameter(false, IQueryBuilder::PARAM_BOOL))
+			->set('status_timestamp', $qb->createNamedParameter($now, IQueryBuilder::PARAM_INT))
+			->where($qb->expr()->lte('status_timestamp', $qb->createNamedParameter($olderThan, IQueryBuilder::PARAM_INT)))
+			->andWhere($qb->expr()->orX(
+				$qb->expr()->eq('is_user_defined', $qb->createNamedParameter(false, IQueryBuilder::PARAM_BOOL)),
```
Comment (ChristophWurst) [inline apps/user_status/lib/Db/UserStatusMapper.php:141]:
```suggestion
				$qb->expr()->eq('is_user_defined', $qb->createNamedParameter(false, IQueryBuilder::PARAM_BOOL), IQueryBuilder::PARAM_BOOL),
```

`eq` takes an optional type as well

Comment (georgehrke) [reply apps/user_status/lib/Db/UserStatusMapper.php]: Done

**Review by ChristophWurst — APPROVED** [review]

**Comment by georgehrke** [comment]
Squashed fixup commits

**Comment by georgehrke** [comment]
There is one PR related unit test failure. Will look into it tomorrow

**Comment by georgehrke** [comment]
Squashed commits

**Pull request merged**